### PR TITLE
perf: add cache for hostserviceCreatorWithCache

### PR DESCRIPTION
**Pull request by ansyral**
when compared pre-release docfx with last release docfx, the perf shows that Load phase takes much more time. Add cache to fix it.

@vwxyzh @vicancy @hellosnow @qinezh @superyyrrzz @chenkennt

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs [hunk 1]
```diff
@@ -111,5 +116,29 @@ private void PostCreate(HostService hostService, IEnumerable<FileAndType> files)
                 IncrementalContext.ReportModelLoadInfo(hostService, loadedFiles, BuildPhase.PreBuildBuild);
             }
         }
+
+        private bool ShouldProcessorTraceInfoCore(IDocumentProcessor processor)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs]: >ShouldProcessorTraceInfoCore [](start = 21, length = 28)

question: why the method is checking incremental, but naming with should processor trace info?

**Comment by superyyrrzz** [comment]
:clock1:

File: src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs (hunk 1 above)
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs]: if a processor, like resourceProcessor, doesn't support incremental, we won't save any intermediate files for it, which is the name's meaning.

**Comment by qinezh** [comment]
:shipit:

**Comment by ansyral** [comment]
close it as it turns out this isn't the root cause that Load becomes so much slower

**Comment by ansyral** [comment]
after not calling gitutility in ResourceDocumentProcessor.Load, the perf is back to normal. still open this as it could improve perf when files count is large

File: src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs [hunk 2]
```diff
@@ -27,26 +30,28 @@ public HostServiceCreatorWithIncremental(DocumentBuildContext context) : base(co
 
         public override bool ShouldProcessorTraceInfo(IDocumentProcessor processor)
         {
-            if (!(processor is ISupportIncrementalDocumentProcessor))
+            var key = $"trace-{processor.Name}";
+            bool result;
+            if (_cache.TryGetValue(key, out result))
```
Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs]: nice

File: src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs [hunk 3]
```diff
@@ -111,5 +116,29 @@ private void PostCreate(HostService hostService, IEnumerable<FileAndType> files)
                 IncrementalContext.ReportModelLoadInfo(hostService, loadedFiles, BuildPhase.PreBuildBuild);
             }
         }
+
+        private bool ShouldProcessorTraceInfoCore(IDocumentProcessor processor)
+        {
+            if (!(processor is ISupportIncrementalDocumentProcessor))
+            {
+                Logger.LogVerbose($"Processor {processor.Name} cannot suppport incremental build because the processor doesn't implement {nameof(ISupportIncrementalDocumentProcessor)} interface.");
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/HostServiceCreatorWithIncremental.cs]: update the suppport to support? :)

---
In reply to: [91652735](https://github.com/dotnet/docfx/pull/982#discussion_r91652735) [](ancestors = 91652735)

**Comment by hellosnow** [comment]
:shipit:

**Pull request merged**